lib/collection/src/collection_manager/optimizers/mod.rs

```diff
@@ -1,5 +1,177 @@
+    ///
+    /// Always keeps the last `KEEP_TRACKERS` trackers.
+    fn truncate(&mut self) {
+        let truncate_range = self.descriptions.len().saturating_sub(KEEP_LAST_TRACKERS);
+
+        // Find items to truncate, start removing from the back
+        let truncate = self
+            .descriptions
+            .iter()
+            .enumerate()
+            .take(truncate_range)
+            .filter(|(_, tracker)| match tracker.state.lock().status {
+                TrackerStatus::Optimizing | TrackerStatus::Error(_) => false,
+                TrackerStatus::Done | TrackerStatus::Cancelled(_) => true,
+            })
+            .map(|(index, _)| index)
+            .collect::<Vec<_>>();
+        truncate.into_iter().rev().for_each(|index| {
+            self.descriptions.remove(index);
+        });
+    }
+
+    /// Convert log into list of objects usable in telemetry
+    pub fn to_telemetry(&self) -> Vec<TrackerTelemetry> {
+        self.descriptions
+            .iter()
+            // Show latest items first
+            .rev()
+            .map(Tracker::to_telemetry)
+            .collect()
+    }
+}
+
+/// Tracks the state of an optimizer
+#[derive(Clone, Debug)]
+pub struct Tracker {
+    /// Name of the optimizer
+    pub name: String,
+    /// Segment IDs being optimized
+    pub segment_ids: Vec<SegmentId>,
+    /// Start time of the optimizer
+    pub start_at: DateTime<Utc>,
+    /// Latest state of the optimizer
+    pub state: Arc<Mutex<TrackerState>>,
+}
+
+impl Tracker {
+    /// Start a new optimizer tracker
+    pub fn start(name: impl Into<String>, segment_ids: Vec<SegmentId>) -> Self {
+        Self {
+            name: name.into(),
+            segment_ids,
+            state: Default::default(),
+            start_at: Utc::now(),
+        }
+    }
+
+    /// Get handle to this tracker, allows updating state
+    pub fn handle(&self) -> TrackerHandle {
+        self.state.clone().into()
+    }
+
+    /// Convert into object used in telemetry
+    pub fn to_telemetry(&self) -> TrackerTelemetry {
+        let state = self.state.lock();
+        TrackerTelemetry {
+            name: self.name.clone(),
+            segment_ids: self.segment_ids.clone(),
+            status: state.status.clone(),
+            start_at: self.start_at,
+            end_at: state.end_at,
+        }
+    }
+}
+
+/// Tracker object used in telemetry
+#[derive(Serialize, Deserialize, Clone, Debug, JsonSchema)]
+pub struct TrackerTelemetry {
+    /// Name of the optimizer
+    pub name: String,
+    /// Segment IDs being optimized
+    pub segment_ids: Vec<SegmentId>,
+    /// Latest status of the optimizer
+    pub status: TrackerStatus,
+    /// Start time of the optimizer
+    pub start_at: DateTime<Utc>,
+    /// End time of the optimizer
+    pub end_at: Option<DateTime<Utc>>,
+}
+
+/// Handle to an optimizer tracker, allows updating its state
+#[derive(Clone)]
+pub struct TrackerHandle {
+    handle: Arc<Mutex<TrackerState>>,
```

Comment:
It's a performance consideration. An RwLock only makes sense when there are many concurrent reads. Mutex'es are fast, that is because the primitive is simpler.

It will only sporadically lock, so I'm not afraid of a huge number of locks (or reads) happening:
- locks when optimizer starts
- locks when optimizer finishes
- locks when fetching telemetry
- if new tracker is added and history is full, some items may be locked to clean up old entries

All locks are very short.

Reply:
I am mostly thinking about the case where several users would hammer the `telemetry` endpoint.
Mostly due to an error on their side, could it then create contention at this point?

Reply:
You'd have to send a humongous amount of requests to make this a bottleneck.

And if they're hammering it due to issues, there likely are a lot bigger problems elsewhere :)

Reply:
By the way, happy to change it to a RwLock as I feel like it doesn't really matter. It's not hot code anyway.

Reply:
Let's keep it like that, it is easy to change if it ever becomes a problem.
Thank you for the details :+1: